File: packages/ios-sdk/Sources/PaaqIntelligence/PaaqApiClient.swift
import Foundation
#if canImport(UIKit)
import UIKit
#endif

private let baseURL = "https://mookyonwpovxscsbqwwl.supabase.co/functions/v1"
private let sdkVersion = "1.0.0"

/// Real device metadata reported once at init — first-party OS APIs only,
/// nothing inferred or backfilled server-side (unlike the web SDK, which has
/// only a raw User-Agent string to work with).
struct PaaqDeviceMetadata: Encodable {
    let osName: String
    let osVersion: String
    let deviceModel: String
    let deviceType: String
    let screenWidth: Int?
    let screenHeight: Int?

    static func collect() -> PaaqDeviceMetadata {
#if canImport(UIKit)
        let device = UIDevice.current
        let bounds = UIScreen.main.bounds
        let scale = UIScreen.main.scale
        return PaaqDeviceMetadata(
            osName: device.systemName,
            osVersion: device.systemVersion,
            deviceModel: device.model,
            deviceType: device.userInterfaceIdiom == .pad ? "tablet" : "mobile",
            screenWidth: Int(bounds.width * scale),
            screenHeight: Int(bounds.height * scale)
        )
#else
        return PaaqDeviceMetadata(
            osName: "macOS", osVersion: ProcessInfo.processInfo.operatingSystemVersionString,
            deviceModel: "Mac", deviceType: "desktop", screenWidth: nil, screenHeight: nil
        )
#endif
    }
}

private struct SdkInitBody: Encodable {
    let deviceId: String
    let appVersion: String?
    let deviceMetadata: PaaqDeviceMetadata
}

private struct SessionEndBody: Encodable {
    let action = "end"
    let session_id: String
    let duration: Int
    let outcome: String
}

actor PaaqApiClient {
    private let config: PaaqConfig
    private let deviceId: String
    private let session: URLSession

    init(config: PaaqConfig, deviceId: String) {
        self.config = config
        self.deviceId = deviceId
        let sessionConfig = URLSessionConfiguration.default
        sessionConfig.timeoutIntervalForRequest = 10
        self.session = URLSession(configuration: sessionConfig)
    }

    private var headers: [String: String] {
        [
            "Content-Type": "application/json",
            "Authorization": "Bearer \(config.sdkToken)",
            "X-Project-ID": config.projectId,
            "X-SDK-Version": sdkVersion,
            "X-Platform": "ios",
            "X-Environment": config.environment,
        ]
    }

    func sdkInit() async throws -> (sessionId: String?, serverBatchSize: Int?, serverFlushInterval: TimeInterval?) {
        var req = URLRequest(url: URL(string: "\(baseURL)/sdk-init")!)
        req.httpMethod = "POST"
        headers.forEach { req.setValue($1, forHTTPHeaderField: $0) }
        let appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
        req.httpBody = try JSONEncoder().encode(SdkInitBody(
            deviceId: deviceId,
            appVersion: appVersion,
            deviceMetadata: PaaqDeviceMetadata.collect()
        ))

        let (data, _) = try await session.data(for: req)
        let json = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
        let sessionId = json["sessionId"] as? String
        let cfg = json["config"] as? [String: Any]
        let batchSize = cfg?["batchSize"] as? Int
        let flushSecs = (cfg?["syncIntervalSeconds"] as? Double).map { $0 }
        return (sessionId, batchSize, flushSecs)
    }

    func flush(events: [PaaqEvent]) async {
        guard !events.isEmpty else { return }
        do {
            var req = URLRequest(url: URL(string: "\(baseURL)/events")!)
            req.httpMethod = "POST"
            headers.forEach { req.setValue($1, forHTTPHeaderField: $0) }
            req.httpBody = try JSONEncoder().encode(events)
            _ = try await session.data(for: req)
        } catch {
            // fire-and-forget
        }
    }

    func endSession(sessionId: String, duration: Int, outcome: String) async {
        do {
            var req = URLRequest(url: URL(string: "\(baseURL)/sessions")!)
            req.httpMethod = "POST"
            headers.forEach { req.setValue($1, forHTTPHeaderField: $0) }
            req.httpBody = try JSONEncoder().encode(SessionEndBody(session_id: sessionId, duration: duration, outcome: outcome))
            _ = try await session.data(for: req)
        } catch {
            // fire-and-forget
        }
    }

    /// Uploads one recording chunk (a JPEG screenshot for iOS) to the private
    /// session-recordings bucket. Query-param metadata, raw bytes as the
    /// body — mirrors the web SDK's upload shape.
    func uploadRecordingChunk(sessionId: String, sequence: Int, capturedAtIso: String, bytes: Data, contentType: String) async {
        guard var components = URLComponents(string: "\(baseURL)/session-recording-upload") else { return }
        components.queryItems = [
            URLQueryItem(name: "session_id", value: sessionId),
            URLQueryItem(name: "kind", value: "screenshots"),
            URLQueryItem(name: "sequence", value: String(sequence)),
            URLQueryItem(name: "captured_at", value: capturedAtIso),
        ]
        guard let url = components.url else { return }
        do {
            var req = URLRequest(url: url)
            req.httpMethod = "POST"
            headers.forEach { req.setValue($1, forHTTPHeaderField: $0) }
            req.setValue(contentType, forHTTPHeaderField: "Content-Type")
            req.httpBody = bytes
            _ = try await session.data(for: req)
        } catch {
            // fire-and-forget
        }
    }

    func heartbeat() async {
        do {
            var req = URLRequest(url: URL(string: "\(baseURL)/sdk-heartbeat")!)
            req.httpMethod = "POST"
            headers.forEach { req.setValue($1, forHTTPHeaderField: $0) }
            req.httpBody = try JSONEncoder().encode(["deviceId": deviceId])
            _ = try await session.data(for: req)
        } catch {
            // fire-and-forget
        }
    }
}

File: packages/android-sdk/src/main/kotlin/io/paaq/intelligence/PaaqApiClient.kt
package io.paaq.intelligence

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.OutputStreamWriter
import java.net.HttpURLConnection
import java.net.URL

private const val BASE_URL = "https://mookyonwpovxscsbqwwl.supabase.co/functions/v1"
private const val SDK_VERSION = "1.0.0"

internal class PaaqApiClient(private val config: PaaqConfig, private val deviceId: String) {

    private fun applyHeaders(conn: HttpURLConnection) {
        conn.setRequestProperty("Content-Type", "application/json")
        conn.setRequestProperty("Authorization", "Bearer ${config.sdkToken}")
        conn.setRequestProperty("X-Project-ID", config.projectId)
        conn.setRequestProperty("X-SDK-Version", SDK_VERSION)
        conn.setRequestProperty("X-Platform", "android")
        conn.setRequestProperty("X-Environment", config.environment)
    }

    private fun post(path: String, body: String): String? {
        return try {
            val conn = (URL("$BASE_URL/$path").openConnection() as HttpURLConnection).apply {
                requestMethod = "POST"
                doOutput = true
                connectTimeout = 10_000
                readTimeout = 10_000
                applyHeaders(this)
            }
            OutputStreamWriter(conn.outputStream).use { it.write(body) }
            if (conn.responseCode in 200..299) {
                conn.inputStream.bufferedReader().readText()
            } else null
        } catch (_: Exception) { null }
    }

    data class InitResult(val sessionId: String?, val batchSize: Int?, val flushIntervalSeconds: Long?)

    /**
     * Real device metadata reported once at init — first-party Android APIs
     * only (Build.*, DisplayMetrics), nothing inferred or backfilled
     * server-side.
     */
    data class DeviceMetadata(
        val osVersion: String,
        val deviceModel: String,
        val appVersion: String?,
        val screenWidth: Int?,
        val screenHeight: Int?,
    )

    suspend fun sdkInit(deviceMetadata: DeviceMetadata? = null): InitResult = withContext(Dispatchers.IO) {
        val body = JSONObject().put("deviceId", deviceId).apply {
            if (deviceMetadata != null) {
                put("appVersion", deviceMetadata.appVersion)
                put("deviceMetadata", JSONObject()
                    .put("osName", "Android")
                    .put("osVersion", deviceMetadata.osVersion)
                    .put("deviceModel", deviceMetadata.deviceModel)
                    .put("deviceType", "mobile")
                    .put("screenWidth", deviceMetadata.screenWidth)
                    .put("screenHeight", deviceMetadata.screenHeight))
            }
        }.toString()
        val response = post("sdk-init", body) ?: return@withContext InitResult(null, null, null)
        val json = JSONObject(response)
        val sessionId = if (json.has("sessionId")) json.getString("sessionId") else null
        val cfg = if (json.has("config")) json.getJSONObject("config") else null
        val batchSize = cfg?.optInt("batchSize")
        val flushSecs = cfg?.optLong("syncIntervalSeconds")
        InitResult(sessionId, batchSize, flushSecs)
    }

    suspend fun endSession(sessionId: String, durationSeconds: Long, outcome: String) = withContext(Dispatchers.IO) {
        val body = JSONObject()
            .put("action", "end")
            .put("session_id", sessionId)
            .put("duration", durationSeconds)
            .put("outcome", outcome)
            .toString()
        post("sessions", body)
    }

    suspend fun flush(events: List<PaaqEvent>) = withContext(Dispatchers.IO) {
        if (events.isEmpty()) return@withContext
        val arr = JSONArray()
        events.forEach { e ->
            val obj = JSONObject()
                .put("event_name", e.eventName)
                .put("session_id", e.sessionId)
                .put("timestamp", e.timestamp)
            val props = JSONObject()
            e.properties.forEach { (k, v) -> props.put(k, v) }
            obj.put("properties", props)
            arr.put(obj)
        }
        post("events", arr.toString())
    }

    suspend fun heartbeat() = withContext(Dispatchers.IO) {
        val body = JSONObject().put("deviceId", deviceId).toString()
        post("sdk-heartbeat", body)
    }
}
